### PR TITLE
Ftr 2941 studies view

### LINKED ISSUE #2887: As a student I want to be able to see all my mandatory, state-level optional and school-level optional courses at once
#### Details

 - Indicated as the background color of unfilled circles
 - Mandatory courses are indicated by light violet
 - National-level optional courses are indicated by light blue
 - School-level optional courses are indicated by light green

### LINKED ISSUE #2941: As an upper secondary school student I want to be able to see the courses available for online study in "studies" view
#### Details

 - Shown as a grid, with course numbers in horizontal axis and subjects in vertical axis
 - Show mandatory courses as empty "circles" and completed courses as filled "circles"
 - The mandatory subjects are taken from HOPS form

### LINKED ISSUE #2942: As a student I want to see the number of courses in "Studies" view
#### Original

Opiskelija näkee valittujen pakollisten ja valinnaisten kurssien määrät.

#### Details

 - Show the number of completed (and later selected) courses *in total* per student
 - Show also the number of mandatory completed courses separately
 - Shown before the grid

### LINKED ISSUE #2943: As a student I want to see different kinds of courses visually distinctly in "studies" view
#### Original

Opiskelija näkee erityyppiset kurssit eroteltuna visuaalisesti toisistaan sekä selitteen kullekin kurssityypille

### LINKED ISSUE #3076: Legend in "studies" view
Show a legend that tells what the colors and numbers mean in the grid in studies view

### LINKED ISSUE #3099: Studies plan view - All shapes & colors to legend
Put all the shapes and colors to legend

### LINKED ISSUE #3100: Studies plan view - menus for courses
Courses should have menus that will contain more information about the state they're in.

Is done in #2941

#### Details

 - Turn off until there's content for the menus

### LINKED ISSUE #3108: Show study status of courses in studies view
#### Possible states

 - Not enrolled
 - Enrolled
 - Marked incomplete
 - Assessed
 - Failed

### LINKED ISSUE #3109: Move HOPS form as a sub-view of studies view
#### Details

 - The form is now a JSF page in profile view
 - It should be refactored into a JS+DUST+AJAX view as one part of the studies view

### LINKED ISSUE #3127: Show all sciences and humanities in "studies" view

### LINKED ISSUE #3128: Choices made in VOPS form are opt-out instead of opt-in

### LINKED ISSUE #3131: Show studies view only to opted in students, others see just transcript of records
#### Details

### LINKED ISSUE #3151: Finish "studies" view
#### Details

 - Check up that everything is in order for prime time

### PR DESCRIPTION
**Pull request by ilmoeuro**
- Fixes #3151 
 - Fixes #2942 
 - Fixes #3108 
 - Fixes #3128 
 - Fixes #3127 
 - Fixes #3131 
 - Fixes #2887 
 - Fixes #2943 
 - Fixes #3099 
 - Fixes #3076 
 - Fixes #2941 
 - Fixes #3100 
 - Fixes #3109

### CODE REVIEW

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java [hunk 1]
```diff
@@ -1,71 +1,315 @@
...
+  @RESTPermit(handling = Handling.INLINE)
+  public Response getVops(@PathParam("IDENTIFIER") String studentIdentifierString) {
+
+    String educationTypeMappingString = pluginSettingsController.getPluginSetting("transcriptofrecords", "educationTypeMapping");
+    EducationTypeMapping educationTypeMapping = new EducationTypeMapping();
+    if (educationTypeMappingString != null) {
+      try {
+        educationTypeMapping = new ObjectMapper().readValue(educationTypeMappingString, EducationTypeMapping.class);
+      } catch (IOException e) {
+        logger.log(Level.INFO, "Education type mapping not set");
```
Comment (pkukkon) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java]: Should be returned as a bad request as a null educationTypeMapping would result in an NPE later on

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java [hunk 2]
```diff
@@ -1,71 +1,315 @@
...
+            boolean workspaceUserExists = false;
+            for (Workspace workspace : workspaces) {
+              WorkspaceEntity workspaceEntity =
+                  workspaceController.findWorkspaceEntity(workspace);
+              WorkspaceUserEntity workspaceUser =
+                  workspaceUserEntityController.findWorkspaceUserByWorkspaceEntityAndUserIdentifier(
+                      workspaceEntity,
+                      studentIdentifier);
+              SchoolDataIdentifier workspaceIdentifier = new SchoolDataIdentifier(workspace.getIdentifier(), workspace.getSchoolDataSource());
+              workspaceAssessments.addAll(gradingController.listWorkspaceAssessments(workspaceIdentifier, studentIdentifier));
```
Comment (pkukkon) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java]: Even if Pyramus is capable of storing multiple assessments per workspace, only the latest one should be considered the one that actually matters. At least the order of assessments retrieved via gradingController should be checked, and it probably wouldn't hurt to have a method that simply returns the most recent workspace assessment, if any.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java [hunk 3]
```diff
@@ -1,71 +1,315 @@
...
+              if (educationSubtypeIdentifier != null) {
+                break;
+              }
+            }
+            Mandatority mandatority = educationTypeMapping.getMandatority(educationSubtypeIdentifier);
+            CourseCompletionState state = CourseCompletionState.NOT_ENROLLED;
+            if (workspaceUserExists) {
+              state = CourseCompletionState.ENROLLED;
+            }
+            for (WorkspaceAssessment workspaceAssessment : workspaceAssessments) {
```
Comment (pkukkon) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java]: These two for loops can be refactored if dealing with a single workspace assessment to begin with (see above).

Comment (ilmoeuro) [inline muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java]: In addition of there being multiple workspace assessments per workspace, as per above comment, there may be multiple workspaces with the same subject and course number (eg. nonstop and "ryhmäkurssi"). We need to consider the situation where the student has failed one of those courses and passed another.

File: muikku/src/main/webapp/resources/theme-muikku/sass/flex/records.scss [hunk 4]
```diff
@@ -234,4 +600,30 @@
...
+  label {
+    margin: 0;
+    padding: 0;
+  }
+
+  input:not(:first-child) {
+    margin-left: 5px;
+  }
+
+  input[type="text"],
```
Comment (avukkonen) [inline muikku/src/main/webapp/resources/theme-muikku/sass/flex/records.scss]: Not now.

Comment (avukkonen) [inline muikku/src/main/webapp/resources/theme-muikku/sass/flex/records.scss]: Not now

File: muikku/src/main/webapp/resources/theme-muikku/sass/flex/records.scss [hunk 5]
```diff
@@ -9,40 +9,400 @@
...
+.tr-vops-legend-content {
+  border-top: 1px solid $environment-legend-border-color;
+  border-bottom: 1px solid $environment-legend-border-color;
+  cursor: pointer;  
+  max-height: 25px;
+
+  overflow-y: hidden;
+  padding: 5px;
+  transition-duration: .5s;
+  transition-property: all;
```
Comment (avukkonen) [inline muikku/src/main/webapp/resources/theme-muikku/sass/flex/records.scss]: Not now

File: muikku/src/main/webapp/resources/theme-muikku/sass/flex/profile.scss [hunk 6]
```diff
@@ -3,16 +3,16 @@
 @import "styles";
 @import "jquery.datepicker";
 
-#profile { 
-  
+#profile {
```
Comment (avukkonen) [inline muikku/src/main/webapp/resources/theme-muikku/sass/flex/profile.scss]: No.

File: muikku/src/main/webapp/resources/theme-muikku/sass/flex/records.scss [hunk 7]
```diff
@@ -2,47 +2,397 @@
...
+
+.tr-vops-legend-content {
+  border-bottom: 1px solid $environment-legend-border-color;
+  border-top: 1px solid $environment-legend-border-color;
+  cursor: pointer;  
+  max-height: 25px;
+  overflow-y: hidden;
+  padding: 5px;
+  transition-duration: 0.5s;
+  transition-property: all;
```
Comment (avukkonen) [inline muikku/src/main/webapp/resources/theme-muikku/sass/flex/records.scss]: No

**Pull request merged**